### ISSUE TITLE
Errors in `@after_verify` cause compiler to exit

### ISSUE THREAD

**Issue by zachdaniel** [elixir-lang/elixir#13149]
### Elixir and Erlang/OTP versions

Erlang/OTP 26 [erts-14.0.2] [source] [64-bit] [smp:10:10] [ds:10:10:10] [async-threads:1] [jit]

Elixir 1.15.4 (compiled with Erlang/OTP 24)

### Operating system

MacOS

### Current behavior

Reproduction: https://github.com/zachdaniel/after_verify_bug

Relevant code:

```elixir
defmodule Verify do
  def broken() do
    defmodule Foo do
      @after_verify __MODULE__

      def __after_verify__(_) do
        raise "what"
      end
    end
  rescue
    _e ->
      {:error, "something went wrong"}
  end

  def not_broken() do
    defmodule Bar do
      @after_compile __MODULE__

      def __after_compile__(_, _) do
        raise "what"
      end
    end

    :ok
  rescue
    _e ->
      {:error, "something went wrong"}
  end
end
```

These two functions have different behavior. The first will exit, and the second will raise. Ash raises errors in the `@after_verify` hook, and this is causing problems with some language server implementations that are expecting to be able to rescue errors from the compiler.

### Expected behavior

Errors raised in any compilation hook should have the same behavior OR it should be documented that you should not raise errors in the `@after_verify` callback.

**Comment by zachdaniel** [comment]
The makes sense to me :) I don't think there is a good option for us aside from raising in the after_verify, unless we want to convert a whole host of validations that currently fail the compiler (that were moved there as a compile time improvement and to allow for certain kinds of checks to be done, i.e verified routes) into warnings. That is an option, but I'd rather language servers assume the compiler can exit 😆

**Comment by lukaszsamson** [comment]
> I'd rather language servers assume the compiler can exit 😆

That's what ElixirLS does

**Comment by josevalim** [comment]
I don’t think we should document this. Any Elixir process can raise and also terminate due to links. I don’t think we should say “this behaves like any other Elixir process/code”. :)

**Comment by zachdaniel** [comment]
Yeah, it was definitely unexpected. Like the strange part is that there is clearly something different about `after_verify`, i.e that it is run in a different process than the rest and/or doesn't catch the exit in the same way (or something along those lines). Perhaps even documenting your recommendation not to raise in that callback? TBH I'll probably still do it though, sorry 🙇

**Comment by josevalim** [comment]
My point is that it is not about the callback. You just cannot expect that the code won’t raise if a linked process crashes. The after_verify callback is just one way of emulating said failure, because it happens in another process, but not the only one.

About after_verify, I will think more about it more. I would prefer it to not raise but it may also be fine. If I change the docs, I want to do on a concrete reason rather than gut feeling.

**Comment by josevalim** [comment]
But I will push a note that we don’t guarantee on which process after verify runs!
